gwcs/wcs.py

```diff
@@ -485,42 +504,39 @@ def invert(self, *args, **kwargs):
         result : tuple or value
             Returns a tuple of scalar or array values for each axis. Unless
             ``input_frame.naxes == 1`` when it shall return the value.
+            The return type will be `~astropy.unit.Quantity` objects if the
+            transform returns ``Quantity`` objects, else values.
 
         """
         with_units = kwargs.pop('with_units', False)
+        if with_units:
+            raise NotImplementedError("no with_units for you")
 
-        if not utils.isnumerical(args[0]):
-            args = self.output_frame.coordinate_to_quantity(*args)
-            if self.output_frame.naxes == 1:
-                args = [args]
-            try:
-                if not self.backward_transform.uses_quantity:
-                    args = utils.get_values(self.output_frame.unit, *args)
-            except (NotImplementedError, KeyError):
-                args = utils.get_values(self.output_frame.unit, *args)
-
-        if 'with_bounding_box' not in kwargs:
-            kwargs['with_bounding_box'] = True
-
-        if 'fill_value' not in kwargs:
-            kwargs['fill_value'] = np.nan
+        return self._call_backward(*args, **kwargs)
 
+    def _call_backward(self, *args, with_bounding_box=True, fill_value=np.nan, **kwargs):
         try:
+            transform = self.backward_transform
+            # Validate that the input type matches what the transform expects
+            input_is_quantity = any((isinstance(a, u.Quantity) for a in args))
+            if not input_is_quantity and transform.uses_quantity:
+                args = self._add_units_input(args, self.output_frame)
+            if not transform.uses_quantity and input_is_quantity:
+                args = self._remove_units_input(args, self.output_frame)
+
             # remove iterative inverse-specific keyword arguments:
             akwargs = {k: v for k, v in kwargs.items() if k not in _ITER_INV_KWARGS}
-            result = self.backward_transform(*args, **akwargs)
+            result = transform(*args, **akwargs)
         except (NotImplementedError, KeyError):
-            result = self.numerical_inverse(*args, **kwargs, with_units=with_units)
-
-        if with_units and self.input_frame:
-            if self.input_frame.naxes == 1:
-                return self.input_frame.coordinates(result)
-            else:
-                return self.input_frame.coordinates(*result)
+            # Always strip units for numerical inverse
+            args = self._remove_units_input(args, self.output_frame)
+            result = self.numerical_inverse(*args, **kwargs)
         else:
             return result
 
-    def numerical_inverse(self, *args, **kwargs):
+    def numerical_inverse(self, *args, tolerance=1e-5, maxiter=50, adaptive=True,
```

Comment:
I have modified this to be numerical only, and then plan to make `invert` handle high-level objects.